### DISCUSSION TITLE
I2C doesn't work

### DISCUSSION THREAD

**Discussion by MariaA2005** [micropython/micropython#10661] [RP2040 / Pico]
I have a MPU9265 connected to the pico on the pins GPIO 8 (sda) and 9 (scl). The original code for the MPU returned Err5, so I tried to scan the devices. It returns an empty array (code follows): 
`from machine import I2C, Pin

#i2c = I2C(0)
i2c = I2C(0, scl=Pin(9), sda=Pin(8), freq=40000)
devices = i2c.scan()
print(devices)`

I have tried with other pins (0,1/16,17), different versions of micropython (v.14, v.19, and  v1.19.1 on 2022-08-31, which is the one I am using right now), to put `time.sleep(1)` between the declaration of i2c and the next line and to declare sda and scl as `sda = Pin(16, Pin.PULL_UP)

scl = Pin(17, Pin.PULL_UP)`. I have checked the connections, tried it on two different picos. The wires and breadboard work perfectly fine for another device with UART connections; however, I have tried a BMP 180, which also uses i2c, and the issue is the same.

**Comment by peterhinch** [comment]
I assume you're aware that the numbers on the Pico PCB don't correspond to GPIO numbers.

Evidently the device isn't being detected. It's hard to be sure what's going on - one option might be to try [SoftI2C](http://docs.micropython.org/en/latest/library/machine.I2C.html#machine.SoftI2C). Another thing to try is supplementing the internal pullups with external resistors connected to 3.3V as the internal pullups are rather weak. Anything in the range 2.2KΩ to 10KΩ would be fine.

**Reply by cve2022** [reply]
> I assume you're aware that the numbers on the Pico PCB don't correspond to GPIO numbers.

I think you meant the physical pin numbers. AFAICT Maria's code is  OK as far as I2C 0 configuration is concerned.

> 
> Evidently the device isn't being detected. It's hard to be sure what's going on - one option might be to try [SoftI2C](http://docs.micropython.org/en/latest/library/machine.I2C.html#machine.SoftI2C). Another thing to try is supplementing the internal pullups with external resistors connected to 3.3V as the internal pullups are rather weak. Anything in the range 2.2KΩ to 10KΩ would be fine.

I agree. The internal weak pull-ups can make the I2C bus too slow.  In my experience, 10K works fine.

@MariaA2005 , assuming the resistors are missing, if it doesn't work after adding them, try reducing the frequency to 100000.
